text/0000-extension-layer.md

```diff
@@ -0,0 +1,90 @@
+# Meta
+[meta]: #meta
+- Name: Add extension layer to exchange data
+- Start Date: 2023-10-09
+- Author(s): [c0d1ngm0nk3y](https://github.com/c0d1ngm0nk3y), [pbusko](https://github.com/pbusko)
+- Status: Draft <!-- Acceptable values: Draft, Approved, On Hold, Superseded -->
+- RFC Pull Request:
+- CNB Pull Request:
+- CNB Issue:
+- Related: [RFC#0105 Support Dockerfiles](https://github.com/buildpacks/rfcs/blob/main/text/0105-dockerfiles.md)
+- Supersedes: N/A
+
+# Summary
+[summary]: #summary
+
+This RFC introduces support for Extension Layers to allow data transfer between the build environment and the Kaniko execution.
+
+# Motivation
+[motivation]: #motivation
+
+This change allows extensions to possess their own layers to utilize during the generation/extend process. Additionally, it ensures that extension output does not inadvertently interfere with other extension or buildpack layers during the build, and it does not unintentionally become part of the final application image.
+
+This would allow distroless run images to be extended.
+
+# What it is
+[what-it-is]: #what-it-is
+
+This follows up on RFC-0105 and proposes that during the execution of the extension's `bin/generate`, an extension is allowed to write arbitrary data to its exclusive layer. This data then becomes accessible during the execution of the `extend` phase via Kaniko context. The content of these extension-specific layers is ignored at build and launch time, it serves only the extension phase.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+Before execution of the `bin/generate`, the lifecycle will create a distinct writable layer for each extension which passed detection. The extensions can then write to these layers, and the Kaniko context is set to the corresponding layer during the `extend` phase instead of the `<app>` directory. The Extension Layers will not be included in the final image by the lifecycle.
+
+The location of the Extension Layer will be provided to the `bin/generate` via additional environment variable `CNB_EXT_LAYER_DIR`.
```

Comment:
Where should the root directory reside within the build container? Thinking of something like `/layers-ext` for the root and `/layers-ext/<extension-id>` for each extension. As extensions and buildpacks share the same id, we would need a new root directory to avoid collisions. We could then simply not mount `/layers-ext` to the build or export containers to avoid polluting the application image.

Reply:
Worth noting, today we provide `CNB_OUTPUT_DIR` which is where we expect to find build.Dockerfiles and run.Dockerfiles (currently, just a temporary directory). Is there any value in combining this directory with the "layers" directory?

Reply:
I think we should consider combining these - `CNB_OUTPUT_DIR` could have `$CNB_OUTPUT_DIR/build-layer` or similar?

Reply:
> I think we should consider combining these - `CNB_OUTPUT_DIR` could have `$CNB_OUTPUT_DIR/build-layer` or similar?

This could be combined with the suggestion from @natalieparellano. We could have `/layers-ext` root directory and the `CNB_OUTPUT_DIR` would point to `/layers-ext/<extension-id>`. Lifecycle would expect run/build Dockerfiles in the root of this directory, as well as `build-layer` folder.